crates/ruma-events/src/room/join_rules.rs

```diff
@@ -1,91 +1,160 @@
 //! Types for the *m.room.join_rules* event.
 
-#[cfg(feature = "unstable-pre-spec")]
-use std::collections::BTreeMap;
-
 use ruma_events_macros::EventContent;
 #[cfg(feature = "unstable-pre-spec")]
 use ruma_identifiers::RoomId;
-use ruma_serde::StringEnum;
 #[cfg(feature = "unstable-pre-spec")]
-use serde::de::{DeserializeOwned, Deserializer, Error};
-use serde::{Deserialize, Serialize};
+use serde::de::DeserializeOwned;
+use serde::{
+    de::{Deserializer, Error},
+    Deserialize, Serialize,
+};
+use serde_json::value::RawValue as RawJsonValue;
 #[cfg(feature = "unstable-pre-spec")]
-use serde_json::{value::RawValue as RawJsonValue, Value as JsonValue};
+use serde_json::Value as JsonValue;
+use std::borrow::Cow;
+#[cfg(feature = "unstable-pre-spec")]
+use std::collections::BTreeMap;
 
 use crate::StateEvent;
 
 /// Describes how users are allowed to join the room.
 pub type JoinRulesEvent = StateEvent<JoinRulesEventContent>;
 
 /// The payload for `JoinRulesEvent`.
-#[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
+#[derive(Clone, Debug, Serialize, EventContent)]
 #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
 #[ruma_event(type = "m.room.join_rules", kind = State)]
 pub struct JoinRulesEventContent {
     /// The type of rules used for users wishing to join this room.
     #[ruma_event(skip_redaction)]
+    #[serde(flatten)]
     pub join_rule: JoinRule,
-
-    /// Allow rules used for the `restricted` join rule.
-    #[cfg(feature = "unstable-pre-spec")]
-    #[serde(default)]
-    #[ruma_event(skip_redaction)]
-    pub allow: Vec<AllowRule>,
 }
 
 impl JoinRulesEventContent {
     /// Creates a new `JoinRulesEventContent` with the given rule.
     pub fn new(join_rule: JoinRule) -> Self {
-        Self {
-            join_rule,
-            #[cfg(feature = "unstable-pre-spec")]
-            allow: Vec::new(),
-        }
+        Self { join_rule }
     }
 
     /// Creates a new `JoinRulesEventContent` with the restricted rule and the given set of allow
     /// rules.
     #[cfg(feature = "unstable-pre-spec")]
     pub fn restricted(allow: Vec<AllowRule>) -> Self {
-        Self { join_rule: JoinRule::Restricted, allow }
+        Self { join_rule: JoinRule::Restricted(Restricted::new(allow)) }
+    }
+}
+
+impl<'de> Deserialize<'de> for JoinRulesEventContent {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let join_rule = JoinRule::deserialize(deserializer)?;
```

Comment:
FWIW with there not being any additional fields this impl could actually be derived (using `#[serde(transparent)]` on the struct), but might make sense to keep it as-is, as a reminder that the derive doesn't work that well here.

Reply:
It would also mean less changes needed in case extra fields do need to get added.